### PR TITLE
feat: DeepSeekMoE

### PR DESCRIPTION
**Pull request by llllvvuu**
# What does this PR do?

Upstream custom code from https://huggingface.co/deepseek-ai/deepseek-moe-16b-base/blob/main/modeling_deepseek.py to huggingface/transformers. This is not DeepSeek V2. The newly released DeepSeek-Prover-V1.5 runs on this architecture for example (though without MoE layers, so it is actually just Llama).

https://huggingface.co/models?other=deepseek

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- text models: @ArthurZucker

### CODE REVIEW

**Review by ArthurZucker — COMMENTED** [review]
Hey! feel free to ping me once this is ready fro review!

**Comment by tmm1** [comment]
friendly ping

File: docs/source/en/model_doc/deepseek.md [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+## DeepSeekMoE
+
+### License
+
+The use of DeepSeekMoE models is subject to the [Model License](https://github.com/deepseek-ai/DeepSeek-MoE/blob/main/LICENSE-MODEL). DeepSeekMoE supports commercial use.
+
+### Description
+
+DeepSeekMoE 16B is a Mixture-of-Experts (MoE) language model with 16.4B parameters. It employs an innovative MoE architecture, which involves two principal strategies: fine-grained expert segmentation and shared experts isolation. It is trained from scratch on 2T English and Chinese tokens, and exhibits comparable performance with DeepSeek 7B and LLaMA2 7B, with only about 40% of computations. For research purposes, we release the model checkpoints of DeepSeekMoE 16B Base and DeepSeekMoE 16B Chat to the public, which can be deployed on a single GPU with 40GB of memory without the need for quantization.
+
```
Comment (ArthurZucker) [inline docs/source/en/model_doc/deepseek.md:26]: maybe missing one line about "this model was contributed by" with your HF username!

File: src/transformers/models/deepseek/configuration_deepseek.py [hunk 2]
```diff
@@ -0,0 +1,205 @@
+from transformers.configuration_utils import PretrainedConfig
+from transformers.utils import logging
+
+
+logger = logging.get_logger(__name__)
+
+DEEPSEEK_PRETRAINED_CONFIG_ARCHIVE_MAP = {}
```
Comment (ArthurZucker) [inline src/transformers/models/deepseek/configuration_deepseek.py:7]:
```suggestion
```

File: src/transformers/models/deepseek/configuration_deepseek.py [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+            Number of shared experts, None means dense model.
+        n_routed_experts (`int`, *optional*):
+            Number of routed experts, None means dense model.
+        num_experts_per_tok (`int`, *optional*):
+            Number of selected experts, None means dense model.
+        moe_layer_freq (`int`, *optional*, defaults to 1):
+            The frequency of the MoE layer: one expert layer for every `moe_layer_freq - 1` dense layers.
+        first_k_dense_replace (`int`, *optional*, defaults to 0):
+            Number of dense layers in shallow layers(embed->dense->dense->...->dense->moe->moe...->lm_head).
+                                                            \--k dense layers--/
```
Comment (ArthurZucker) [inline src/transformers/models/deepseek/configuration_deepseek.py:54]: would be nice to have standardization with the other MoE in the library! we usually call this the `sparse_step`: 
https://github.com/huggingface/transformers/blob/6cffc90dcacca489f785fa4689addbc49f8dd17f/src/transformers/models/qwen2_moe/configuration_qwen2_moe.py#L77-L78

File: src/transformers/models/deepseek/configuration_deepseek.py [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+            Padding token id.
+        bos_token_id (`int`, *optional*, defaults to 100000):
+            Beginning of stream token id.
+        eos_token_id (`int`, *optional*, defaults to 100001):
+            End of stream token id.
+        pretraining_tp (`int`, *optional*, defaults to 1):
+            Experimental feature. Tensor parallelism rank used during pretraining. Please refer to [this
+            document](https://huggingface.co/docs/transformers/parallelism) to understand more about it. This value is
+            necessary to ensure exact reproducibility of the pretraining results. Please refer to [this
+            issue](https://github.com/pytorch/pytorch/issues/76232).
```
Comment (ArthurZucker) [inline src/transformers/models/deepseek/configuration_deepseek.py:84]: should not be used!

File: src/transformers/models/deepseek/configuration_deepseek.py [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+                f"got {self.rope_scaling}"
+            )
+        rope_scaling_type = self.rope_scaling.get("type", None)
+        rope_scaling_factor = self.rope_scaling.get("factor", None)
+        if rope_scaling_type is None or rope_scaling_type not in ["linear", "dynamic"]:
+            raise ValueError(
+                f"`rope_scaling`'s type field must be one of ['linear', 'dynamic'], got {rope_scaling_type}"
+            )
+        if rope_scaling_factor is None or not isinstance(rope_scaling_factor, float) or rope_scaling_factor <= 1.0:
+            raise ValueError(f"`rope_scaling`'s factor field must be a float > 1, got {rope_scaling_factor}")
```
Comment (ArthurZucker) [inline src/transformers/models/deepseek/configuration_deepseek.py:205]:
```suggestion
```

File: src/transformers/models/deepseek/configuration_deepseek.py [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+
+        self.num_key_value_heads = num_key_value_heads
+        self.hidden_act = hidden_act
+        self.initializer_range = initializer_range
+        self.rms_norm_eps = rms_norm_eps
+        self.pretraining_tp = pretraining_tp
+        self.use_cache = use_cache
+        self.rope_theta = rope_theta
+        self.rope_scaling = rope_scaling
+        self._rope_scaling_validation()
```
Comment (ArthurZucker) [inline src/transformers/models/deepseek/configuration_deepseek.py:174]:
```suggestion
        rope_config_validation(self)
```

File: src/transformers/models/deepseek/configuration_deepseek.py [hunk 7]
```diff
@@ -0,0 +1,205 @@
+from transformers.configuration_utils import PretrainedConfig
+from transformers.utils import logging
```
Comment (ArthurZucker) [inline src/transformers/models/deepseek/configuration_deepseek.py:2]:
```suggestion
from ...configuration_utils import PretrainedConfig
from ...utils import logging
from ...modeling_rope_utils import rope_config_validation
```

File: tests/generation/test_utils.py
Comment (ArthurZucker) [inline tests/generation/test_utils.py:1]: good catch thanks for fixing

File: tests/models/musicgen/test_modeling_musicgen.py
Comment (ArthurZucker) [inline tests/models/musicgen/test_modeling_musicgen.py:1]: looks unrelated

File: tests/models/deepseek/test_modeling_deepseek.py [hunk 8]
```diff
@@ -0,0 +1,613 @@
...
+                    dummy_input,
+                    attention_mask=dummy_attention_mask,
+                    max_new_tokens=max_new_tokens,
+                    do_sample=False,
+                    use_cache=True,
+                )
+
+
+@slow
+@require_torch_gpu  # NotImplementedError: The operator 'aten::scatter_reduce.two_out' is not currently implemented for the MPS device.
```
Comment (ArthurZucker) [inline tests/models/deepseek/test_modeling_deepseek.py:534]: if there is a gate, you might need the `require_read_token` wrapper!

File: src/transformers/models/deepseek/modeling_deepseek.py [hunk 9]
```diff
@@ -0,0 +1,1754 @@
+# coding=utf-8
+# Copyright 2023 DeepSeek-AI and The HuggingFace Inc. team. All rights reserved.
```
Comment (ArthurZucker) [inline src/transformers/models/deepseek/modeling_deepseek.py:2]:
```suggestion
# Copyright 2024 DeepSeek-AI and The HuggingFace Inc. team. All rights reserved.
```

File: src/transformers/models/deepseek/modeling_deepseek.py [hunk 10]
```diff
@@ -0,0 +1,1754 @@
...
+#
+# This code is based on EleutherAI's GPT-NeoX library and the GPT-NeoX
+# and OPT implementations in this library. It has been modified from its
+# original forms to accommodate minor architectural differences compared
+# to GPT-NeoX and OPT used by the Meta AI team that trained the model.
```
Comment (ArthurZucker) [inline src/transformers/models/deepseek/modeling_deepseek.py:7]:
```suggestion
```

File: src/transformers/models/deepseek/modeling_deepseek.py [hunk 11]
```diff
@@ -0,0 +1,1754 @@
...
+                [F.linear(x, gate_proj_slices[i]) for i in range(self.config.pretraining_tp)], dim=-1
+            )
+            up_proj = torch.cat([F.linear(x, up_proj_slices[i]) for i in range(self.config.pretraining_tp)], dim=-1)
+
+            intermediate_states = (self.act_fn(gate_proj) * up_proj).split(slice, dim=2)
+            down_proj = [
+                F.linear(intermediate_states[i], down_proj_slices[i]) for i in range(self.config.pretraining_tp)
+            ]
+            down_proj = sum(down_proj)
+        else:
```
Comment (ArthurZucker) [inline src/transformers/models/deepseek/modeling_deepseek.py:292]:
```suggestion
```
here copied from gemma should be fine

File: src/transformers/models/deepseek/modeling_deepseek.py [hunk 12]
```diff
@@ -0,0 +1,1754 @@
...
+            down_proj = self.down_proj(self.act_fn(self.gate_proj(x)) * self.up_proj(x))
+
+        return down_proj
+
+
+class MoEGate(nn.Module):
```
Comment (ArthurZucker) [inline src/transformers/models/deepseek/modeling_deepseek.py:298]:
```suggestion
class MoeGate(nn.Module):
```

File: src/transformers/models/deepseek/modeling_deepseek.py [hunk 13]
```diff
@@ -0,0 +1,1754 @@
...
+        self.norm_topk_prob = config.norm_topk_prob
+        self.gating_dim = config.hidden_size
+        self.weight = nn.Parameter(torch.empty((self.n_routed_experts, self.gating_dim)))
+        self.reset_parameters()
+
+    def reset_parameters(self) -> None:
+        import torch.nn.init as init
+
+        init.kaiming_uniform_(self.weight, a=math.sqrt(5))
+
```
Comment (ArthurZucker) [inline src/transformers/models/deepseek/modeling_deepseek.py:319]: initis do not happend at the module level in `transformers` but they do happen at the `PretrainedModel` le vel in the `_init_weights` function!

File: src/transformers/models/deepseek/modeling_deepseek.py [hunk 14]
```diff
@@ -0,0 +1,1754 @@
...
+    def forward(self, hidden_states):
+        bsz, seq_len, h = hidden_states.shape
+        ### compute gating score
+        hidden_states = hidden_states.view(-1, h)
+        logits = F.linear(hidden_states, self.weight, None)
+        if self.scoring_func == "softmax":
+            scores = logits.softmax(dim=-1)
+        else:
+            raise NotImplementedError(f"insupportable scoring function for MoE gating: {self.scoring_func}")
```
Comment (ArthurZucker) [inline src/transformers/models/deepseek/modeling_deepseek.py:328]: let's just remove this parameter and just use softmax 😉

File: src/transformers/models/deepseek/modeling_deepseek.py [hunk 15]
```diff
@@ -0,0 +1,1754 @@
...
+
+        ### select top-k experts
+        topk_weight, topk_idx = torch.topk(scores, k=self.top_k, dim=-1, sorted=False)
+
+        ### norm gate to sum 1
+        if self.top_k > 1 and self.norm_topk_prob:
```
Comment (ArthurZucker) [inline src/transformers/models/deepseek/modeling_deepseek.py:334]: does the released model use norm tokp prob or not? We tend to avoid code pathes and would rather fit to what the released model uses! (if this is set to false let's remove it!)

File: src/transformers/models/deepseek/modeling_deepseek.py [hunk 16]
```diff
@@ -0,0 +1,1754 @@
...
+                ).div_(seq_len * aux_topk / self.n_routed_experts)
+                aux_loss = (ce * scores_for_seq_aux.mean(dim=1)).sum(dim=1).mean() * self.alpha
+            else:
+                mask_ce = F.one_hot(topk_idx_for_aux_loss.view(-1), num_classes=self.n_routed_experts)
+                ce = mask_ce.float().mean(0)
+                Pi = scores_for_aux.mean(0)
+                fi = ce * self.n_routed_experts
+                aux_loss = (Pi * fi).sum() * self.alpha
+        else:
+            aux_loss = None
```
Comment (ArthurZucker) [inline src/transformers/models/deepseek/modeling_deepseek.py:358]: Mmmm similarly we usually try to isolate the loss, see `modeling_mixtral` for an example. This gives more freedom to the user and we just have a default `aux_loss` function that can be easily overwritten!

File: src/transformers/models/deepseek/modeling_deepseek.py [hunk 17]
```diff
@@ -0,0 +1,1754 @@
...
+        ctx.dtype = loss.dtype
+        ctx.required_aux_loss = loss.requires_grad
+        return x
+
+    @staticmethod
+    def backward(ctx, grad_output):
+        grad_loss = None
+        if ctx.required_aux_loss:
+            grad_loss = torch.ones(1, dtype=ctx.dtype, device=grad_output.device)
+        return grad_output, grad_loss
```
Comment (ArthurZucker) [inline src/transformers/models/deepseek/modeling_deepseek.py:380]: let's remove this!

File: src/transformers/models/deepseek/modeling_deepseek.py [hunk 18]
```diff
@@ -0,0 +1,1754 @@
...
+            y = torch.empty_like(hidden_states)
+            for i, expert in enumerate(self.experts):
+                y[flat_topk_idx == i] = expert(hidden_states[flat_topk_idx == i])
+            y = (y.view(*topk_weight.shape, -1) * topk_weight.unsqueeze(-1)).sum(dim=1)
+            y = y.view(*orig_shape)
+            y = AddAuxiliaryLoss.apply(y, aux_loss)
+        else:
+            y = self.moe_infer(hidden_states, flat_topk_idx, topk_weight.view(-1, 1)).view(*orig_shape)
+        if self.config.n_shared_experts is not None:
+            y = y + self.shared_experts(identity)
```
Comment (ArthurZucker) [inline src/transformers/models/deepseek/modeling_deepseek.py:420]: same comment about the loss also let's avoid single name variable!

File: src/transformers/models/deepseek/modeling_deepseek.py [hunk 19]
```diff
@@ -0,0 +1,1754 @@
...
+            start_idx = 0 if i == 0 else tokens_per_expert[i - 1]
+            if start_idx == end_idx:
+                continue
+            expert = self.experts[i]
+            exp_token_idx = token_idxs[start_idx:end_idx]
+            expert_tokens = x[exp_token_idx]
+            expert_out = expert(expert_tokens)
+            expert_out.mul_(flat_expert_weights[idxs[start_idx:end_idx]])
+            expert_cache.scatter_reduce_(0, exp_token_idx.view(-1, 1).repeat(1, x.shape[-1]), expert_out, reduce="sum")
+        return expert_cache
```
Comment (ArthurZucker) [inline src/transformers/models/deepseek/modeling_deepseek.py:439]: I don't really. mind having 2 forwards, one for inference one for training, tho this is not super common

File: src/transformers/models/deepseek/modeling_deepseek.py [hunk 20]
```diff
@@ -0,0 +1,1754 @@
...
+            hidden_states = hidden_states.repeat_interleave(self.num_experts_per_tok, dim=0)
+            y = torch.empty_like(hidden_states)
+            for i, expert in enumerate(self.experts):
+                y[flat_topk_idx == i] = expert(hidden_states[flat_topk_idx == i])
+            y = (y.view(*topk_weight.shape, -1) * topk_weight.unsqueeze(-1)).sum(dim=1)
+            y = y.view(*orig_shape)
+            y = AddAuxiliaryLoss.apply(y, aux_loss)
+        else:
+            y = self.moe_infer(hidden_states, flat_topk_idx, topk_weight.view(-1, 1)).view(*orig_shape)
+        if self.config.n_shared_experts is not None:
```
Comment (ArthurZucker) [inline src/transformers/models/deepseek/modeling_deepseek.py:419]: let's avoid code path if the model uses shared experts, let's add this otherwise let's remove!

File: src/transformers/models/deepseek/modeling_deepseek.py [hunk 21]
```diff
@@ -0,0 +1,1754 @@
...
+    num_key_value_heads, seqlen, head_dim) to (batch, num_attention_heads, seqlen, head_dim)
+    """
+    batch, num_key_value_heads, slen, head_dim = hidden_states.shape
+    if n_rep == 1:
+        return hidden_states
+    hidden_states = hidden_states[:, :, None, :, :].expand(batch, num_key_value_heads, n_rep, slen, head_dim)
+    return hidden_states.reshape(batch, num_key_value_heads * n_rep, slen, head_dim)
+
+
+# Copied from transformers.models.llama.modeling_llama.LlamaAttention with Llama->Deepseek
```
Comment (ArthurZucker) [inline src/transformers/models/deepseek/modeling_deepseek.py:455]: we should rather copy from Gemma, removes the need for pretaining tps!

File: src/transformers/models/deepseek/modeling_deepseek.py [hunk 22]
```diff
@@ -0,0 +1,1754 @@
...
+
+        if position_embeddings is None:
+            logger.warning_once(
+                "The attention layers in this model are transitioning from computing the RoPE embeddings internally "
+                "through `position_ids` (2D tensor with the indexes of the tokens), to using externally computed "
+                "`position_embeddings` (Tuple of tensors, containing cos and sin). In v4.45 `position_ids` will be "
+                "removed and `position_embeddings` will be mandatory."
+            )
+            cos, sin = self.rotary_emb(value_states, position_ids)
+        else:
```
Comment (ArthurZucker) [inline src/transformers/models/deepseek/modeling_deepseek.py:542]: as this is a new model, it will not support out of the box legacy behaviours

File: src/transformers/models/deepseek/modeling_deepseek.py [hunk 23]
```diff
@@ -0,0 +1,1754 @@
...
+
+        self.mlp = (
+            DeepseekMoE(config)
+            if (
+                config.n_routed_experts is not None
+                and layer_idx >= config.first_k_dense_replace
+                and layer_idx % config.moe_layer_freq == 0
+            )
+            else DeepseekMLP(config)
+        )
```
Comment (ArthurZucker) [inline src/transformers/models/deepseek/modeling_deepseek.py:831]: let's split in 2 lines

**Review by ArthurZucker — COMMENTED** [review]
Thanks for the PR! I am not 100 sure I caught what are the architectural differences with say QwenMoe which also has the shared experts piece! 

Also let's try to match the architecture of the release models, removing unecessary codepathes!

**Comment by wavy-jung** [comment]
Any update?

**Comment by llllvvuu** [comment]
> Any update?

Sorry, I got busy, will get back to this

**Comment by ArthurZucker** [comment]
still nice to have! 🤗
